# 2020-01-19

Doing some more exploratory data analysis on comment data. 

Want to see if I can discover the similarity between two different YT videos

In [1]:
import pandas as pd

In [2]:
import os

In [3]:
os.getcwd()

'C:\\Users\\Ronald Maj\\Documents\\GitHub\\InsightDataProject\\notebooks'

In [4]:
cd ..

C:\Users\Ronald Maj\Documents\GitHub\InsightDataProject


In [5]:
drew1_df = pd.read_csv("data/raw/drew1.csv")

In [6]:
drew1_df.head(7)

CommID authorDisplayName  \
0                         UgyMAptldbaotq25pgx4AaABAg        tomatoanus   
1  UgyMAptldbaotq25pgx4AaABAg.8kLILEv0s_h93xcOkRbfQL  Nathaniel Huerta   
2  UgyMAptldbaotq25pgx4AaABAg.8kLILEv0s_h93sUhWOu88m    Abe Abi Haydar   
3  UgyMAptldbaotq25pgx4AaABAg.8kLILEv0s_h93eZplH5ug7          Yummyboi   
4  UgyMAptldbaotq25pgx4AaABAg.8kLILEv0s_h93_dNCV8tK1          Jan Wick   
5  UgyMAptldbaotq25pgx4AaABAg.8kLILEv0s_h93JojN1C7qc            qwerty   
6                         Ugwga9AWIlVnajO3RR94AaABAg       Wendy McCoy   

                               authorProfileImageUrl  \
0  https://yt3.ggpht.com/a/AGF-l7-dWGEdE94iYyfghH...   
1  https://yt3.ggpht.com/a/AGF-l78sghDmpMleaviF44...   
2  https://yt3.ggpht.com/a/AGF-l7-p_uJ-NlFNC1Sbjm...   
3  https://yt3.ggpht.com/a/AGF-l7_yW6h1RXGK0MIe6F...   
4  https://yt3.ggpht.com/a/AGF-l7--SwayUtPbOOD1Q0...   
5  https://yt3.ggpht.com/a/AGF-l79llzvO-jCC7m5WLO...   
6  https://yt3.ggpht.com/a/AGF-l7_lxlfY6eiEHLrWNg...   

                                    authorChannelUrl  \
0  http://www.youtube.com/channel/UCm6GSA5OROHcIB...   
1  http://www.youtube.com/channel/UC3hksSESHg3dDJ...   
2  http://www.youtube.com/channel/UCNib--KiuE0TaG...   
3  http://www.youtube.com/channel/UC_nuyvsPqhiHuF...   
4  http://www.youtube.com/channel/UCZgWKtUWiHHK5f...   
5  http://www.youtube.com/channel/UCC0Uq-HTbCA7uU...   
6  http://www.youtube.com/channel/UCsPwH-Fb4ffPK2...   

                         authorChannelId      videoId  \
0  {'value': 'UCm6GSA5OROHcIBNkXkH53zQ'}  rSLWFA_glbk   
1  {'value': 'UC3hksSESHg3dDJmiK6kPaHQ'}  rSLWFA_glbk   
2  {'value': 'UCNib--KiuE0TaGqIpUW9_sA'}  rSLWFA_glbk   
3  {'value': 'UC_nuyvsPqhiHuF2X2XlELGg'}  rSLWFA_glbk   
4  {'value': 'UCZgWKtUWiHHK5fz-AK4lK6w'}  rSLWFA_glbk   
5  {'value': 'UCC0Uq-HTbCA7uUnHsGV6V8A'}  rSLWFA_glbk   
6  {'value': 'UCsPwH-Fb4ffPK2Upamn65wQ'}  rSLWFA_glbk   

                                         textDisplay  \
0  he was just eatin some eggs outside a girls wi...   
1                                He was stalking her   
2                    @Pramey Sontakke stfu like stap   
3                              literally everything🤣   
4                                          mmmmm egg   
5                            ITS ON A THURSDAY NIGHT   
6  Apparently pedophiles can only either grunt or...   

                                        textOriginal  canRate viewerRating  \
0  he was just eatin some eggs outside a girls wi...     True         none   
1                                He was stalking her     True         none   
2                    @Pramey Sontakke stfu like stap     True         none   
3                              literally everything🤣     True         none   
4                                          mmmmm egg     True         none   
5                            ITS ON A THURSDAY NIGHT     True         none   
6  Apparently pedophiles can only either grunt or...     True         none   

   likeCount               publishedAt                 updatedAt  \
0      12447  2018-08-24T18:28:29.000Z  2018-08-24T18:28:29.000Z   
1          0  2020-01-19T03:30:44.000Z  2020-01-19T03:30:44.000Z   
2          0  2020-01-17T03:38:35.000Z  2020-01-17T03:38:35.000Z   
3          0  2020-01-11T17:54:02.000Z  2020-01-11T17:54:02.000Z   
4          0  2020-01-09T19:57:29.000Z  2020-01-09T19:57:29.000Z   
5          0  2020-01-03T07:09:41.000Z  2020-01-03T07:09:41.000Z   
6       2079  2019-07-14T23:48:01.000Z  2019-07-14T23:48:01.000Z   

                     parentId  
0                           0  
1  UgyMAptldbaotq25pgx4AaABAg  
2  UgyMAptldbaotq25pgx4AaABAg  
3  UgyMAptldbaotq25pgx4AaABAg  
4  UgyMAptldbaotq25pgx4AaABAg  
5  UgyMAptldbaotq25pgx4AaABAg  
6                           0

In [7]:
len(drew1_df)

1626

In [8]:
import scripts.yt_cm as cm

In [9]:
# Initially ran:
# drew2_df = cm.create_comments_df('iI48g7YTZFA')
drew2_df = pd.read_csv("data/raw/drew2.csv")

In [10]:
len(drew2_df)

1986

In [11]:
def inner_val(st):
    return st['value']

In [12]:
drew2_df['authorChannelId'].map(inner_val)

TypeError: string indices must be integers

Having problems with the conversion of the commenters unique id value to a string only (i.e. not a dict) format. 

I have decided to go down the route of a psychographic market segmentation based on YouTube comments.
The idea being that taking the first 100 comments from the (say) 30 latest videos of a channel would be give a good overall picture of the audience that watches that channel (assuming a number of things about the comments of course - no trolls, no bots, no artificial inflation of numbers, etc.)

To keep the assumption at least partially true, I have tried to stick to channels that are non-political and have a vlog / talk to the camera style of production. This way the topics in each video hopefully will be varied enough that audience responses will also be quite varied and a good idea of what the interests of the audience actually is. Of course there will still be certain topics that the channel may concentrate on, but that may be useful for matching a new video to.

The main idea is to try to use the Five-factor personality traits to put channels on a sliding scale of personality types. Categorise channel audiences according to these personality traits and therefore make it easier for marketers to target the audience they are looking for when advertising their products.

# 2020-01-20

Anyway, what I will do now is test out whether there are some simple (off-the-shelf) metrics I can use to show the similarity between two channels. Firstly, I will just start with analysis of a handful of videos.
The two Drew Gooden (commentary / comedy) videos I have already downloaded, plus another third one from a seemingly distant video - say Jenn Im (beauty, travel, lifestyle).

In [13]:
# Initially needed to save:
# drew2_df.to_csv("data/raw/drew2.csv")

In [21]:
# Intially run:
# jen1 = cm.create_comments_df('ZcCp6ghTP78', order_option='relevance', max_pgs=2)
# jen1.to_csv('data/raw/jen1.csv')
jen1_df = pd.read_csv('data/raw/jen1.csv')

In [15]:
drew1_wholetext = ''
for text in drew1_df[drew1_df['parentId'] == '0']['textDisplay']:
    drew1_wholetext = drew1_wholetext + '. ' + text

In [16]:
drew1_wholetext[0:2000]

'. he was just eatin some eggs outside a girls window, whats wrong with that. Apparently pedophiles can only either grunt or sound like a gangster from Looney Tunes. I learn so much from these videos.. I thought I recognized you. Subscribed for months and I didn\'t notice you were the road work ahead guy.. *"It\'s past 2:00 AM, and you\'re standing outside a little girls window eating EgGs."*. Let me think of some other age then 14\n\n*She told me she was ten*. Man: *sniffs hair*\nAlso man: *grunts*\n\nEveryone: What\'s weird about that?. 13:08 that censor was on point 👌. “Why is there a chair INFRONT of the door?!?”\nThat’s the chair Jackie used when you got kidnapped by Logan *pail* right before *Denny* jumps out the window and goes to the *suicide hospital*. Creepy online person: Hey are you 14?\nNormal answer: Oh no, I\'m some other age.... *I mean, I know it\'s August but I can\'t wait to see YouTube rewind with the only youtuber in the world*. Currently saving up for a Road Work 

In [17]:
print(drew1_wholetext[0:2000])

. he was just eatin some eggs outside a girls window, whats wrong with that. Apparently pedophiles can only either grunt or sound like a gangster from Looney Tunes. I learn so much from these videos.. I thought I recognized you. Subscribed for months and I didn't notice you were the road work ahead guy.. *"It's past 2:00 AM, and you're standing outside a little girls window eating EgGs."*. Let me think of some other age then 14

*She told me she was ten*. Man: *sniffs hair*
Also man: *grunts*

Everyone: What's weird about that?. 13:08 that censor was on point 👌. “Why is there a chair INFRONT of the door?!?”
That’s the chair Jackie used when you got kidnapped by Logan *pail* right before *Denny* jumps out the window and goes to the *suicide hospital*. Creepy online person: Hey are you 14?
Normal answer: Oh no, I'm some other age.... *I mean, I know it's August but I can't wait to see YouTube rewind with the only youtuber in the world*. Currently saving up for a Road Work Ahead sign. No,

### Test

Test to see if the comment thread from the first half of the text matches the second.

Need to tokenize and then run a similarity measure on it. I have mostly used the following guide to help me out here:
https://kanoki.org/2018/12/27/text-matching-cosine-similarity/
https://towardsdatascience.com/benchmarking-python-nlp-tokenizers-3ac4735100c5

The idea here being that we can use the tools from sklearn to do a number of tokenization / vectorization and even run cosine similarity on the text documents.

In [18]:
from nltk.tokenize.regexp import regexp_tokenize
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import TfidfVectorizer

In [19]:
# Adapted from: https://towardsdatascience.com/benchmarking-python-nlp-tokenizers-3ac4735100c5

# Remove non-words:
def clean_text(text):
    # replace new line and return with space
    text = text.replace("\n", " ").replace("\r"," ")
    # replace punctuation marks with space
    punc_list = '!"#$%&()*+,-./:;<=>?@[\]{}|^_~' + '0123456789'
    t=str.maketrans(dict.fromkeys(punc_list, " "))
    text = text.translate(t)
    # replace single quote with empty character
    t = str.maketrans(dict.fromkeys("'`",""))
    text = text.translate(t)
    return text

# Create a list of words:
def renltk_tokenize(text):
    text= clean_text(text)
    # search for any white space and tokenize as word using regex '\s+' 
    words = regexp_tokenize(text, pattern = '\s+', gaps=True)
    return words

In [28]:
# Create whole text of the comments from the YouTube videos:

drew1_wholetext = ''
for text in drew1_df[drew1_df['parentId'] == '0']['textDisplay']:
    drew1_wholetext = drew1_wholetext + '. ' + text
    
drew2_wholetext = ''
for text in drew2_df[drew2_df['parentId'] == '0']['textDisplay']:
    drew2_wholetext = drew2_wholetext + ' ' + text    

jen1_wholetext = ''
for text in jen1_df[jen1_df['parentId'] == '0']['textDisplay']:
    jen1_wholetext = jen1_wholetext + ' ' + text 

In [29]:
# Split up the comments sections into two halves for each video in order to get a test sample
# Should expect most similarity between the two halves of the same drew video, less so between two drew videos
# and even less between the jen videos and the drew videos:

half1 = int(len(drew1_wholetext)/2)
half2 = int(len(drew2_wholetext)/2)

jhalf1 = int(len(jen1_wholetext)/2)

drew1_part1 = drew1_wholetext[0:half1]
drew1_part2 = drew1_wholetext[half1:]
drew2_part1 = drew2_wholetext[0:half2]
drew2_part2 = drew2_wholetext[half2:]
jen1_part1 = jen1_wholetext[0:jhalf1]
jen1_part2 = jen1_wholetext[jhalf1:] 

In [34]:
# Adapted from https://kanoki.org/2018/12/27/text-matching-cosine-similarity/

corpus = [drew1_part1, drew1_part2, drew2_part1, drew2_part2, jen1_part1, jen1_part2]

vectorizer = TfidfVectorizer()
trsfm=vectorizer.fit_transform(corpus)
#pd.DataFrame(trsfm.toarray(),columns=vectorizer.get_feature_names(),index=['Drew1-1','Drew1-2','Drew2-1','Drew2-2'])
parts_name = ['Drew1-1','Drew1-2','Drew2-1','Drew2-2','Jen1-1','Jen1-2']

pd.DataFrame(cosine_similarity(trsfm, trsfm),index=parts_name,columns=parts_name)

Drew1-1   Drew1-2   Drew2-1   Drew2-2    Jen1-1    Jen1-2
Drew1-1  1.000000  0.933291  0.832886  0.850158  0.691575  0.721108
Drew1-2  0.933291  1.000000  0.820286  0.847365  0.701766  0.735160
Drew2-1  0.832886  0.820286  1.000000  0.963130  0.683302  0.715230
Drew2-2  0.850158  0.847365  0.963130  1.000000  0.699679  0.739808
Jen1-1   0.691575  0.701766  0.683302  0.699679  1.000000  0.789228
Jen1-2   0.721108  0.735160  0.715230  0.739808  0.789228  1.000000

As can be seen above, the similarity between the Drew videos is high, but that of the Jen parts with each other is quite low.

The Drew videos have many more comments than the Jenn videos, so therefore it would be a good idea to check what would happen if we restrict ourselves to the case of top 100 / 200 comments:

In [35]:
# Create one text string of the top 100 comments from the YouTube videos:

drew1_wholetext = ''
for text in drew1_df[drew1_df['parentId'] == '0']['textDisplay'].head(100):
    drew1_wholetext = drew1_wholetext + '. ' + text
    
drew2_wholetext = ''
for text in drew2_df[drew2_df['parentId'] == '0']['textDisplay'].head(100):
    drew2_wholetext = drew2_wholetext + ' ' + text    

jen1_wholetext = ''
for text in jen1_df[jen1_df['parentId'] == '0']['textDisplay'].head(100):
    jen1_wholetext = jen1_wholetext + ' ' + text 
    
half1 = int(len(drew1_wholetext)/2)
half2 = int(len(drew2_wholetext)/2)

jhalf1 = int(len(jen1_wholetext)/2)

drew1_part1 = drew1_wholetext[0:half1]
drew1_part2 = drew1_wholetext[half1:]
drew2_part1 = drew2_wholetext[0:half2]
drew2_part2 = drew2_wholetext[half2:]
jen1_part1 = jen1_wholetext[0:jhalf1]
jen1_part2 = jen1_wholetext[jhalf1:] 

corpus = [drew1_part1, drew1_part2, drew2_part1, drew2_part2, jen1_part1, jen1_part2]

vectorizer = TfidfVectorizer()
trsfm=vectorizer.fit_transform(corpus)
parts_name = ['Drew1-1','Drew1-2','Drew2-1','Drew2-2','Jen1-1','Jen1-2']

pd.DataFrame(cosine_similarity(trsfm, trsfm),index=parts_name,columns=parts_name)

Drew1-1   Drew1-2   Drew2-1   Drew2-2    Jen1-1    Jen1-2
Drew1-1  1.000000  0.595144  0.489256  0.490112  0.407508  0.420691
Drew1-2  0.595144  1.000000  0.532776  0.537677  0.465025  0.484303
Drew2-1  0.489256  0.532776  1.000000  0.731451  0.465339  0.487702
Drew2-2  0.490112  0.537677  0.731451  1.000000  0.499045  0.516758
Jen1-1   0.407508  0.465025  0.465339  0.499045  1.000000  0.689129
Jen1-2   0.420691  0.484303  0.487702  0.516758  0.689129  1.000000

Results are a lot closer now. Perhaps we can try top 200:

In [36]:
# Create one text string of the top 200 comments from the YouTube videos:

drew1_wholetext = ''
for text in drew1_df[drew1_df['parentId'] == '0']['textDisplay'].head(200):
    drew1_wholetext = drew1_wholetext + '. ' + text
    
drew2_wholetext = ''
for text in drew2_df[drew2_df['parentId'] == '0']['textDisplay'].head(200):
    drew2_wholetext = drew2_wholetext + ' ' + text    

jen1_wholetext = ''
for text in jen1_df[jen1_df['parentId'] == '0']['textDisplay'].head(200):
    jen1_wholetext = jen1_wholetext + ' ' + text 
    
half1 = int(len(drew1_wholetext)/2)
half2 = int(len(drew2_wholetext)/2)

jhalf1 = int(len(jen1_wholetext)/2)

drew1_part1 = drew1_wholetext[0:half1]
drew1_part2 = drew1_wholetext[half1:]
drew2_part1 = drew2_wholetext[0:half2]
drew2_part2 = drew2_wholetext[half2:]
jen1_part1 = jen1_wholetext[0:jhalf1]
jen1_part2 = jen1_wholetext[jhalf1:] 

corpus = [drew1_part1, drew1_part2, drew2_part1, drew2_part2, jen1_part1, jen1_part2]

vectorizer = TfidfVectorizer()
trsfm=vectorizer.fit_transform(corpus)
parts_name = ['Drew1-1','Drew1-2','Drew2-1','Drew2-2','Jen1-1','Jen1-2']

pd.DataFrame(cosine_similarity(trsfm, trsfm),index=parts_name,columns=parts_name)

Drew1-1   Drew1-2   Drew2-1   Drew2-2    Jen1-1    Jen1-2
Drew1-1  1.000000  0.742457  0.626346  0.637959  0.557621  0.597398
Drew1-2  0.742457  1.000000  0.632211  0.648912  0.594968  0.631254
Drew2-1  0.626346  0.632211  1.000000  0.861286  0.582219  0.607821
Drew2-2  0.637959  0.648912  0.861286  1.000000  0.597217  0.626545
Jen1-1   0.557621  0.594968  0.582219  0.597217  1.000000  0.769085
Jen1-2   0.597398  0.631254  0.607821  0.626545  0.769085  1.000000

Definitely better with 200 comments, however the similarity between Drew1-2 and Drew2-2 is very similar to that of Jen1-2 (0.632 vs 0.631). There may in fact be some similarity in those specific comments, however it may be difficult to check.

Perhaps I can try the Count Vectorizer and see how it compares to the tf-idf:

In [37]:
# 
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()
trsfm=vectorizer.fit_transform(corpus)
parts_name = ['Drew1-1','Drew1-2','Drew2-1','Drew2-2','Jen1-1','Jen1-2']

pd.DataFrame(cosine_similarity(trsfm, trsfm),index=parts_name,columns=parts_name)

Drew1-1   Drew1-2   Drew2-1   Drew2-2    Jen1-1    Jen1-2
Drew1-1  1.000000  0.859251  0.789245  0.797220  0.721472  0.760506
Drew1-2  0.859251  1.000000  0.769063  0.786341  0.749186  0.782423
Drew2-1  0.789245  0.769063  1.000000  0.920646  0.724755  0.746906
Drew2-2  0.797220  0.786341  0.920646  1.000000  0.738409  0.766729
Jen1-1   0.721472  0.749186  0.724755  0.738409  1.000000  0.875223
Jen1-2   0.760506  0.782423  0.746906  0.766729  0.875223  1.000000

Similarity still remains quite high. Therefore I will stick with tf-idf for now.

I need to develop the code for automatic video download given a YouTube video.

Ideas:
 - Get a list of YouTube channel ids (based on the list I have curated of vloggers)
 - Run the ids through a function to find the 30 latest videos from that creator (ids)
 - Run the existing code to create a database of comment threads

